metadata/src/from_into/v14.rs

```diff
@@ -308,69 +310,116 @@ impl ExtrinsicPartTypeIds {
 fn generate_outer_enums(
     metadata: &mut v14::RuntimeMetadataV14,
 ) -> v15::OuterEnums<scale_info::form::PortableForm> {
-    let call_enum = metadata
-        .types
-        .types
-        .iter()
-        .find(|ty| {
+    let mut path = None;
+
+    let mut find_type = |name: &str| {
+        metadata.types.types.iter().find_map(|ty| {
             let Some(ident) = ty.ty.path.ident() else {
-                return false;
+                return None;
             };
-            ident == "RuntimeCall"
-        })
-        .expect("RuntimeCall exists in V14; qed");
 
-    let event_enum = metadata
-        .types
-        .types
-        .iter()
-        .find(|ty| {
-            let Some(ident) = ty.ty.path.ident() else {
-                return false;
+            if ident != name {
+                return None;
+            }
+
+            let TypeDef::Variant(_) = &ty.ty.type_def else {
+                return None;
             };
-            ident == "RuntimeEvent"
+
+            if path.is_none() {
+                let mut segments = ty.ty.path.segments.clone();
+                segments.pop();
+                path = Some(segments);
+            }
+
+            Some(ty.id)
         })
-        .expect("RuntimeEvent exists in V14; qed");
+    };
 
-    let call_ty = call_enum.id.into();
-    let event_ty = event_enum.id.into();
+    let call_enum = find_type("RuntimeCall").or_else(|| find_type("Call"));
+    let event_enum = find_type("RuntimeEvent").or_else(|| find_type("Event"));
+    let error_enum = find_type("RuntimeError").or_else(|| find_type("Error"));
```

Comment:
One issue: `RuntimeCall`, `RuntimeEvent` and `RuntimeError` are fairly unique names which are unlikely to be used for random other types. but `Call`, `Event` and `Error` are not, and in fact all of these names are used for other enums in the type registry (each pallet could have one or more of these types declared).

Is the `Call` enum that you found actually one of those per-pallet ones? I have a feeling though that possibly an old version of Substrate used `Call` rather than `RuntimeCall` or something like that, but can't remember for sure.

Anyway, in the generated polkakdot metadata, the types we want (`RuntimeCall` etc) are all at paths like `runtime_types::polkadot_runtime::RuntimeCall`, `runtime_types::polkadot_runtime::RuntimeError` and `runtime_types::polkadot_runtime::RuntimeEvent`. (other types all live in `runtime_types` but in other places). I wonder whether the `_runtime` part of the `polkadot_runtime` path can be used to help find these things. Or else, maybe we just return an error and don't try falling back to anything else.

Reply:
Yep that makes sense, checking if a `Call` is indeed the enum we are looking for can become quite tedious :D

I've changed the code to not have fallbacks; propagate the error back to the user and only generate the `RuntimeError` if not present, which seems to be more straight forward